core/src/mindustry/editor/MapEditorDialog.java

```diff
@@ -665,19 +666,34 @@ private void tryExit(){
         ui.showConfirm("@confirm", "@editor.unsaved", this::hide);
     }
 
-    private void addBlockSelection(Table table){
-        Table content = new Table();
-        pane = new ScrollPane(content);
+    private void addBlockSelection(Table cont){
+        blockSelection = new Table();
+        pane = new ScrollPane(blockSelection);
         pane.setFadeScrollBars(false);
         pane.setOverscroll(true, false);
         pane.exited(() -> {
             if(pane.hasScroll()){
                 Core.scene.setScrollFocus(view);
             }
         });
-        ButtonGroup<ImageButton> group = new ButtonGroup<>();
 
-        int i = 0;
+        cont.table(search -> {
+            search.image(Icon.zoom).padRight(8);
+            search.field("", this::rebuildBlockSelection)
+            .name("editor/search").maxTextLength(maxNameLength).get().setMessageText("@players.search");
+        }).pad(5);
+        cont.row();
+        cont.table(Tex.underline, extra -> extra.labelWrap(() -> editor.drawBlock.localizedName).width(200f).center()).growX();
+        cont.row();
+        cont.add(pane).expandY().top().left();
+
+        rebuildBlockSelection("");
+    }
+
+    private void rebuildBlockSelection(String searchText){
+        blockSelection.clear();
+
+        Seq<Block> filteredBlocks = new Seq<>();
```

Comment:
Instead of a list, how about this:

- instead of `filteredBlocks.isEmpty()`, check `i == 0`
- instead of `filteredBlocks.first();`, just set `drawBlock` to the current when `i == 0` and you're iterating

Reply:
Good, resolved in https://github.com/Anuken/Mindustry/pull/3265/commits/255d1c511ef07c6e7cbb9cf3353f3ff616026e62.